prometheusremotewriteexporter logs spurious errors with WAL configured and no metrics to be sent

### Component(s)

exporter/prometheusremotewrite

### What happened?

## Description

When the prometheusremotewriteexporter is configured to use a WAL directory and is receiving no metrics to export, it will emit error logs that `error processing WAL entries` where there are, in fact, none to be processed. The exporter should fall back to using `fsnotify.NewWatcher()` but it does not. 

## Steps to Reproduce

run otelcol-contrib with included config

## Expected Result

no errors and no error logs

## Actual Result

error logs emitted every 5s


### Collector version

v0.81.0

### Environment information

## Environment
OS: osx, colima, docker
docker run --rm --mount type=bind,source=$(pwd),target=/etc/otelcol/   otel/opentelemetry-collector-contrib:latest --config /etc/otelcol/config.yaml

### OpenTelemetry Collector configuration

```yaml
exporters:
  prometheusremotewrite:
    endpoint: "http://host.docker.internal:9090/api/v1/push"
    wal:
      directory: /etc/otelcol/
extensions:
  health_check: {}
  memory_ballast:
    size_in_percentage: 40
processors:
  batch: {}
  memory_limiter:
    check_interval: 5s
    limit_percentage: 80
    spike_limit_percentage: 25

receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318
service:
  extensions:
  - health_check
  - memory_ballast
  pipelines:
    metrics:
      exporters:
      - prometheusremotewrite
      processors:
      - memory_limiter
      - batch
      receivers:
      - otlp
```


### Log output

```shell
2023-07-19T23:00:48.424Z        info    service/service.go:158  Everything is ready. Begin running and processing data.
2023-07-19T23:00:52.535Z        error   prw.wal prometheusremotewriteexporter@v0.77.0/wal.go:182        error processing WAL entries    {"kind": "exporter", "data_type": "metrics", "name": "prometheusremotewrite", "error": "not found"}
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter.(*prweWAL).run.func1
        github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter@v0.77.0/wal.go:182
2023-07-19T23:00:56.659Z        error   prw.wal prometheusremotewriteexporter@v0.77.0/wal.go:182        error processing WAL entries    {"kind": "exporter", "data_type": "metrics", "name": "prometheusremotewrite", "error": "not found"}
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter.(*prweWAL).run.func1
        github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter@v0.77.0/wal.go:182
2023-07-19T23:01:00.792Z        error   prw.wal prometheusremotewriteexporter@v0.77.0/wal.go:182        error processing WAL entries    {"kind": "exporter", "data_type": "metrics", "name": "prometheusremotewrite", "error": "not found"}
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter.(*prweWAL).run.func1
        github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter@v0.77.0/wal.go:182
```


### Additional context

It appears that the error check in `*prweWAL.readPrompbFromWAL` for `wal.ErrNotFound` is not correctly catching it, as the `"error": "not found"` text is rising up to the `"error processing WAL entries"` error logging. 

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/prometheusremotewriteexporter/wal.go#L358

I've taken some time to look at this today. One of my colleagues tried solving this on #20875, but it got stale, and he moved on to other things. I'm still wrapping my head around the problem, and from the previous PR I can see discussions around deadlocks.

After updating the WAL mutexes with [delock](https://github.com/ietxaniz/delock) and running the exporter, I got this:

```
./bin/otelcontribcol_darwin_arm64 --config config.yml 
2025-01-31T18:03:53.231-0300    info    service@v0.118.1-0.20250131104636-a737a48402e0/service.go:186   Setting up own telemetry...
2025-01-31T18:03:53.232-0300    info    service@v0.118.1-0.20250131104636-a737a48402e0/service.go:252   Starting otelcontribcol...      {"Version": "0.118.0-dev", "NumCPU": 12}
2025-01-31T18:03:53.232-0300    info    extensions/extensions.go:39     Starting extensions...
2025-01-31T18:03:53.233-0300    info    otlpreceiver@v0.118.1-0.20250131104636-a737a48402e0/otlp.go:112 Starting GRPC server    {"kind": "receiver", "name": "otlp", "data_type": "metrics", "endpoint": "localhost:4317"}
2025-01-31T18:03:53.233-0300    info    otlpreceiver@v0.118.1-0.20250131104636-a737a48402e0/otlp.go:169 Starting HTTP server    {"kind": "receiver", "name": "otlp", "data_type": "metrics", "endpoint": "localhost:4318"}
2025-01-31T18:03:53.234-0300    info    prometheusreceiver@v0.118.0/metrics_receiver.go:118     Starting discovery manager      {"kind": "receiver", "name": "prometheus", "data_type": "metrics"}
2025-01-31T18:03:53.236-0300    info    targetallocator/manager.go:184  Scrape job added        {"kind": "receiver", "name": "prometheus", "data_type": "metrics", "jobName": "otel-collector"}
2025-01-31T18:03:53.236-0300    info    service@v0.118.1-0.20250131104636-a737a48402e0/service.go:275   Everything is ready. Begin running and processing data.
2025-01-31T18:03:53.236-0300    info    prometheusreceiver@v0.118.0/metrics_receiver.go:187     Starting scrape manager {"kind": "receiver", "name": "prometheus", "data_type": "metrics"}
panic: 

        Deadlock detected




        * * * * * WRITE LOCK - 1 occurrences * * * * * * * * 
                github.com/ietxaniz/delock@v0.3.0/mutex.go:58 +0x3c
                github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter@v0.118.0/wal.go:336 +0x4c
                github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter@v0.118.0/wal.go:231 +0x15c
                github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter@v0.118.0/wal.go:174 +0xdc
                github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter@v0.118.0/wal.go:164 +0x1f4





goroutine 52 [running]:
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter.(*prweWAL).persistToWAL(0x140015e0480, {0x1400115a460, 0x1, 0x0?})
        github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter@v0.118.0/wal.go:317 +0x270
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter.(*prwExporter).handleExport(0x140015c98c0, {0x10c3b4910, 0x140009e2e00}, 0x14001767890, {0x0, 0x0, 0x0})
        github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter@v0.118.0/exporter.go:257 +0xfc
```

This shows me that the code path `prwe.handleExport -> persistToWAL` is trying to acquire the lock, while it has already been acquired by another code path `prwe.Start -> turnOnWALIfEnabled -> run -> continuallyPopWALThenExport -> readPrompbFromWAL`

I'm working on the PR

--- a/.chloggen/prw-WAL-deadlock.yaml
+++ b/.chloggen/prw-WAL-deadlock.yaml
@@ -1,0 +1,19 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: prometheusremotewriteexproter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: |
+  Resolves a deadlock in the WAL by temporarily releasing a lock while waiting for new writes to the WAL.
+# One or more tracking issues related to the change
+issues: [19363, 24399, 15277]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+
+change_logs: [user]

--- a/exporter/prometheusremotewriteexporter/go.mod
+++ b/exporter/prometheusremotewriteexporter/go.mod
@@ -4,7 +4,6 @@ go 1.22.0
 
 require (
 	github.com/cenkalti/backoff/v4 v4.3.0
-	github.com/fsnotify/fsnotify v1.8.0
 	github.com/gogo/protobuf v1.3.2
 	github.com/golang/snappy v0.0.4
 	github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal v0.118.0
@@ -36,6 +35,8 @@ require (
 	go.uber.org/multierr v1.11.0
 	go.uber.org/zap v1.27.0
 )
+
+require github.com/fsnotify/fsnotify v1.8.0 // indirect
 
 require (
 	github.com/cespare/xxhash/v2 v2.3.0 // indirect

--- a/exporter/prometheusremotewriteexporter/wal.go
+++ b/exporter/prometheusremotewriteexporter/wal.go
@@ -12,18 +12,15 @@ import (
 	"sync/atomic"
 	"time"
 
-	"github.com/fsnotify/fsnotify"
 	"github.com/gogo/protobuf/proto"
 	"github.com/prometheus/prometheus/prompb"
 	"github.com/tidwall/wal"
 	"go.uber.org/multierr"
 	"go.uber.org/zap"
-
-	"github.com/ietxaniz/delock"
 )
 
 type prweWAL struct {
-	mu        delock.Mutex // mu protects the fields below.
+	mu        sync.Mutex // mu protects the fields below.
 	wal       *wal.Log
 	walConfig *WALConfig
 	walPath   string
@@ -32,6 +29,7 @@ type prweWAL struct {
 
 	stopOnce  sync.Once
 	stopChan  chan struct{}
+	rNotify   chan struct{}
 	rWALIndex *atomic.Uint64
 	wWALIndex *atomic.Uint64
 }
@@ -72,6 +70,7 @@ func newWAL(walConfig *WALConfig, exportSink func(context.Context, []*prompb.Wri
 		exportSink: exportSink,
 		walConfig:  walConfig,
 		stopChan:   make(chan struct{}),
+		rNotify:    make(chan struct{}),
 		rWALIndex:  &atomic.Uint64{},
 		wWALIndex:  &atomic.Uint64{},
 	}
@@ -96,11 +95,8 @@ var (
 
 // retrieveWALIndices queries the WriteAheadLog for its current first and last indices.
 func (prwe *prweWAL) retrieveWALIndices() (err error) {
-	lockID, err := prwe.mu.Lock()
-	if err != nil {
-		panic(err)
-	}
-	defer prwe.mu.Unlock(lockID)
+	prwe.mu.Lock()
+	defer prwe.mu.Unlock()
 
 	err = prwe.closeWAL()
 	if err != nil {
@@ -132,11 +128,8 @@ func (prwe *prweWAL) retrieveWALIndices() (err error) {
 func (prwe *prweWAL) stop() error {
 	err := errAlreadyClosed
 	prwe.stopOnce.Do(func() {
-		lockID, err := prwe.mu.Lock()
-	if err != nil {
-		panic(err)
-	}
-	defer prwe.mu.Unlock(lockID)
+		prwe.mu.Lock()
+		defer prwe.mu.Unlock()
 
 		close(prwe.stopChan)
 		err = prwe.closeWAL()
@@ -268,11 +261,8 @@ func (prwe *prweWAL) closeWAL() error {
 }
 
 func (prwe *prweWAL) syncAndTruncateFront() error {
-	lockID, err := prwe.mu.Lock()
-	if err != nil {
-		panic(err)
-	}
-	defer prwe.mu.Unlock(lockID)
+	prwe.mu.Lock()
+	defer prwe.mu.Unlock()
 
 	if prwe.wal == nil {
 		return errNilWAL
@@ -312,11 +302,8 @@ func (prwe *prweWAL) exportThenFrontTruncateWAL(ctx context.Context, reqL []*pro
 // write them to the Write-Ahead-Log so that shutdowns won't lose data, and that the routine that
 // reads from the WAL can then process the previously serialized requests.
 func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
-	lockID, err := prwe.mu.Lock()
-	if err != nil {
-		panic(err)
-	}
-	defer prwe.mu.Unlock(lockID)
+	prwe.mu.Lock()
+	defer prwe.mu.Unlock()
 
 	// Write all the requests to the WAL in a batch.
 	batch := new(wal.Batch)
@@ -329,16 +316,15 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
 		batch.Write(wIndex, protoBlob)
 	}
 
+	// Notify reader go routine that is possibly waiting for writes.
+	select {
+	case prwe.rNotify <- struct{}{}:
+	default:
+	}
 	return prwe.wal.WriteBatch(batch)
 }
 
 func (prwe *prweWAL) readPrompbFromWAL(ctx context.Context, index uint64) (wreq *prompb.WriteRequest, err error) {
-	lockID, err := prwe.mu.Lock()
-	if err != nil {
-		panic(err)
-	}
-	defer prwe.mu.Unlock(lockID)
-
 	var protoBlob []byte
 	for i := 0; i < 12; i++ {
 		// Firstly check if we've been terminated, then exit if so.
@@ -354,10 +340,10 @@ func (prwe *prweWAL) readPrompbFromWAL(ctx context.Context, index uint64) (wreq 
 			index = 1
 		}
 
+		prwe.mu.Lock()
 		if prwe.wal == nil {
 			return nil, fmt.Errorf("attempt to read from closed WAL")
 		}
-
 		protoBlob, err = prwe.wal.Read(index)
 		if err == nil { // The read succeeded.
 			req := new(prompb.WriteRequest)
@@ -368,74 +354,24 @@ func (prwe *prweWAL) readPrompbFromWAL(ctx context.Context, index uint64) (wreq 
 			// Now increment the WAL's read index.
 			prwe.rWALIndex.Add(1)
 
+			prwe.mu.Unlock()
 			return req, nil
+		}
+		prwe.mu.Unlock()
+
+		// If WAL was empty, let's wait for a notification from
+		// the writer go routine.
+		if errors.Is(err, wal.ErrNotFound) {
+			select {
+			case <-prwe.rNotify:
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			}
 		}
 
 		if !errors.Is(err, wal.ErrNotFound) {
 			return nil, err
 		}
-
-		if index <= 1 {
-			// This could be the very first attempted read, so try again, after a small sleep.
-			time.Sleep(time.Duration(1<<i) * time.Millisecond)
-			continue
-		}
-
-		// Otherwise, we couldn't find the record, let's try watching
-		// the WAL file until perhaps there is a write to it.
-		walWatcher, werr := fsnotify.NewWatcher()
-		if werr != nil {
-			return nil, werr
-		}
-		if werr = walWatcher.Add(prwe.walPath); werr != nil {
-			return nil, werr
-		}
-
-		// Watch until perhaps there is a write to the WAL file.
-		watchCh := make(chan error)
-		wErr := err
-		go func() {
-			defer func() {
-				watchCh <- wErr
-				close(watchCh)
-				// Close the file watcher.
-				walWatcher.Close()
-			}()
-
-			select {
-			case <-ctx.Done(): // If the context was cancelled, bail out ASAP.
-				wErr = ctx.Err()
-				return
-
-			case event, ok := <-walWatcher.Events:
-				if !ok {
-					return
-				}
-				switch event.Op {
-				case fsnotify.Remove:
-					// The file got deleted.
-					// TODO: Add capabilities to search for the updated file.
-				case fsnotify.Rename:
-					// Renamed, we don't have information about the renamed file's new name.
-				case fsnotify.Write:
-					// Finally a write, let's try reading again, but after some watch.
-					wErr = nil
-				}
-
-			case eerr, ok := <-walWatcher.Errors:
-				if ok {
-					wErr = eerr
-				}
-			}
-		}()
-
-		if gerr := <-watchCh; gerr != nil {
-			return nil, gerr
-		}
-
-		// Otherwise a write occurred might have occurred,
-		// and we can sleep for a little bit then try again.
-		time.Sleep(time.Duration(1<<i) * time.Millisecond)
 	}
 	return nil, err
 }

--- a/exporter/prometheusremotewriteexporter/wal.go
+++ b/exporter/prometheusremotewriteexporter/wal.go
@@ -18,10 +18,12 @@ import (
 	"github.com/tidwall/wal"
 	"go.uber.org/multierr"
 	"go.uber.org/zap"
+
+	"github.com/ietxaniz/delock"
 )
 
 type prweWAL struct {
-	mu        sync.Mutex // mu protects the fields below.
+	mu        delock.Mutex // mu protects the fields below.
 	wal       *wal.Log
 	walConfig *WALConfig
 	walPath   string
@@ -94,8 +96,11 @@ var (
 
 // retrieveWALIndices queries the WriteAheadLog for its current first and last indices.
 func (prwe *prweWAL) retrieveWALIndices() (err error) {
-	prwe.mu.Lock()
-	defer prwe.mu.Unlock()
+	lockID, err := prwe.mu.Lock()
+	if err != nil {
+		panic(err)
+	}
+	defer prwe.mu.Unlock(lockID)
 
 	err = prwe.closeWAL()
 	if err != nil {
@@ -127,8 +132,11 @@ func (prwe *prweWAL) retrieveWALIndices() (err error) {
 func (prwe *prweWAL) stop() error {
 	err := errAlreadyClosed
 	prwe.stopOnce.Do(func() {
-		prwe.mu.Lock()
-		defer prwe.mu.Unlock()
+		lockID, err := prwe.mu.Lock()
+	if err != nil {
+		panic(err)
+	}
+	defer prwe.mu.Unlock(lockID)
 
 		close(prwe.stopChan)
 		err = prwe.closeWAL()
@@ -260,8 +268,11 @@ func (prwe *prweWAL) closeWAL() error {
 }
 
 func (prwe *prweWAL) syncAndTruncateFront() error {
-	prwe.mu.Lock()
-	defer prwe.mu.Unlock()
+	lockID, err := prwe.mu.Lock()
+	if err != nil {
+		panic(err)
+	}
+	defer prwe.mu.Unlock(lockID)
 
 	if prwe.wal == nil {
 		return errNilWAL
@@ -301,8 +312,11 @@ func (prwe *prweWAL) exportThenFrontTruncateWAL(ctx context.Context, reqL []*pro
 // write them to the Write-Ahead-Log so that shutdowns won't lose data, and that the routine that
 // reads from the WAL can then process the previously serialized requests.
 func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
-	prwe.mu.Lock()
-	defer prwe.mu.Unlock()
+	lockID, err := prwe.mu.Lock()
+	if err != nil {
+		panic(err)
+	}
+	defer prwe.mu.Unlock(lockID)
 
 	// Write all the requests to the WAL in a batch.
 	batch := new(wal.Batch)
@@ -319,8 +333,11 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
 }
 
 func (prwe *prweWAL) readPrompbFromWAL(ctx context.Context, index uint64) (wreq *prompb.WriteRequest, err error) {
-	prwe.mu.Lock()
-	defer prwe.mu.Unlock()
+	lockID, err := prwe.mu.Lock()
+	if err != nil {
+		panic(err)
+	}
+	defer prwe.mu.Unlock(lockID)
 
 	var protoBlob []byte
 	for i := 0; i < 12; i++ {

--- a/exporter/prometheusremotewriteexporter/wal.go
+++ b/exporter/prometheusremotewriteexporter/wal.go
@@ -366,6 +366,8 @@ func (prwe *prweWAL) readPrompbFromWAL(ctx context.Context, index uint64) (wreq 
 			case <-prwe.rNotify:
 			case <-ctx.Done():
 				return nil, ctx.Err()
+			case <-prwe.stopChan:
+				return nil, fmt.Errorf("attempt to read from WAL after stopped")
 			}
 		}
 

--- a/exporter/prometheusremotewriteexporter/wal_test.go
+++ b/exporter/prometheusremotewriteexporter/wal_test.go
@@ -5,13 +5,22 @@ package prometheusremotewriteexporter
 
 import (
 	"context"
+	"io"
+	"net/http"
+	"net/http/httptest"
 	"sort"
 	"testing"
 	"time"
 
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
 	"github.com/prometheus/prometheus/prompb"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/config/confighttp"
+	"go.opentelemetry.io/collector/exporter/exportertest"
 )
 
 func doNothingExportSink(_ context.Context, reqL []*prompb.WriteRequest) error {
@@ -148,4 +157,57 @@ func TestWAL_persist(t *testing.T) {
 	orderByLabelValueForEach(reqLFromWAL)
 	require.Equal(t, reqLFromWAL[0], reqL[0])
 	require.Equal(t, reqLFromWAL[1], reqL[1])
+}
+
+func TestExportWithWALEnabled(t *testing.T) {
+	cfg := &Config{
+		WAL: &WALConfig{
+			Directory: t.TempDir(),
+		},
+		TargetInfo: &TargetInfo{}, // Declared just to avoid nil pointer dereference.
+		CreatedMetric: &CreatedMetric{}, // Declared just to avoid nil pointer dereference.
+	}
+	buildInfo := component.BuildInfo{
+		Description: "OpenTelemetry Collector",
+		Version:     "1.0",
+	}
+	set := exportertest.NewNopSettings()
+	set.BuildInfo = buildInfo
+
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		body, err := io.ReadAll(r.Body)
+		require.NoError(t, err)
+		require.NotNil(t, body)
+		// Receives the http requests and unzip, unmarshalls, and extracts TimeSeries
+		writeReq := &prompb.WriteRequest{}
+		var unzipped []byte
+
+		dest, err := snappy.Decode(unzipped, body)
+		require.NoError(t, err)
+
+		ok := proto.Unmarshal(dest, writeReq)
+		require.NoError(t, ok)
+
+		assert.Len(t, writeReq.Timeseries, 1)
+	}))
+	clientConfig := confighttp.NewDefaultClientConfig()
+	clientConfig.Endpoint = server.URL
+	cfg.ClientConfig = clientConfig
+
+	prwe, err := newPRWExporter(cfg, set)
+	assert.NoError(t, err)
+	assert.NotNil(t, prwe)
+	err = prwe.Start(context.Background(), componenttest.NewNopHost())
+	require.NoError(t, err)
+	assert.NotNil(t, prwe.client)
+
+	metrics := map[string]*prompb.TimeSeries{
+		"test_metric": {
+			Labels:  []prompb.Label{{Name: "__name__", Value: "test_metric"}},
+			Samples: []prompb.Sample{{Value: 1, Timestamp: 100}},
+		},
+	}
+	err = prwe.handleExport(context.Background(), metrics, nil)
+	assert.NoError(t, err)
 }

--- a/exporter/prometheusremotewriteexporter/wal_test.go
+++ b/exporter/prometheusremotewriteexporter/wal_test.go
@@ -209,4 +209,9 @@ func TestExportWithWALEnabled(t *testing.T) {
 	}
 	err = prwe.handleExport(context.Background(), metrics, nil)
 	assert.NoError(t, err)
+
+	// While on Unix systems, t.TempDir() would easily close the WAL files,
+	// on Windows, it doesn't. So we need to close it manually to avoid flaky tests.
+	err = prwe.Shutdown(context.Background())
+	assert.NoError(t, err)
 }

--- a/exporter/prometheusremotewriteexporter/wal_test.go
+++ b/exporter/prometheusremotewriteexporter/wal_test.go
@@ -164,7 +164,7 @@ func TestExportWithWALEnabled(t *testing.T) {
 		WAL: &WALConfig{
 			Directory: t.TempDir(),
 		},
-		TargetInfo: &TargetInfo{}, // Declared just to avoid nil pointer dereference.
+		TargetInfo:    &TargetInfo{},    // Declared just to avoid nil pointer dereference.
 		CreatedMetric: &CreatedMetric{}, // Declared just to avoid nil pointer dereference.
 	}
 	buildInfo := component.BuildInfo{
@@ -174,20 +174,19 @@ func TestExportWithWALEnabled(t *testing.T) {
 	set := exportertest.NewNopSettings()
 	set.BuildInfo = buildInfo
 
-
-	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	server := httptest.NewServer(http.HandlerFunc(func(_ http.ResponseWriter, r *http.Request) {
 		body, err := io.ReadAll(r.Body)
-		require.NoError(t, err)
-		require.NotNil(t, body)
+		assert.NoError(t, err)
+		assert.NotNil(t, body)
 		// Receives the http requests and unzip, unmarshalls, and extracts TimeSeries
 		writeReq := &prompb.WriteRequest{}
 		var unzipped []byte
 
 		dest, err := snappy.Decode(unzipped, body)
-		require.NoError(t, err)
+		assert.NoError(t, err)
 
 		ok := proto.Unmarshal(dest, writeReq)
-		require.NoError(t, ok)
+		assert.NoError(t, ok)
 
 		assert.Len(t, writeReq.Timeseries, 1)
 	}))

